src/sentry/integrations/github/blame.py

```diff
@@ -0,0 +1,201 @@
+            if not ref:
+                logger.info(
+                    f"Branch {ref_name} for repository {full_repo_name} does not exist in GitHub.",
+                    extra=logger_info,
+                )
+                continue
+            for file_path_index, file_path in enumerate(file_paths):
+                blame = ref.get("target", {}).get(f"blame{file_path_index}")
+                if not blame:
+                    logger.info(
+                        f"File {file_path} for branch {ref_name} for repository {full_repo_name} does not exist in GitHub.",
+                        extra=logger_info,
+                    )
+                    continue
+                matching_file, blame_info = _get_matching_file_and_blame(
+                    files=files,
+                    blame_ranges=blame.get("ranges", []),
+                    path=file_path,
+                    repo_name=full_repo_name,
+                    ref=ref_name,
+                )
+                if not blame_info:
+                    logger.info(
+                        f"No matching commit found for line {matching_file.lineno} in file {matching_file.path} in branch {matching_file.ref} in repository {full_repo_name} in GitHub.",
+                        extra=logger_info,
+                    )
+                    continue
+                file_blames.append(blame_info)
+    return file_blames
+
+
+def _get_matching_file_and_blame(
+    files: Sequence[SourceLineInfo],
+    blame_ranges: Sequence[GitHubFileBlameRange],
+    path: str,
+    repo_name: str,
+    ref: str,
+):
+    matching_file = [
+        f for f in files if f.path == path and f.repo.name == repo_name and f.ref == ref
+    ][0]
+    matching_blame_range: GitHubFileBlameRange | None = next(
+        iter(
+            [
+                r
+                for r in blame_ranges
+                if r["startingLine"] <= matching_file.lineno <= r["endingLine"]
+            ]
+        ),
+        None,
+    )
+    if not matching_blame_range:
+        return matching_file, None
+
+    commit = matching_blame_range.get("commit") or {}
+    author = commit.get("author") or {}
+    blame = FileBlameInfo(
+        **asdict(matching_file),
+        commit=CommitInfo(
+            commitId=commit.get("oid"),
+            commitAuthorName=author.get("name"),
+            commitAuthorEmail=author.get("email"),
+            commitMessage=commit.get("message"),
+            committedDate=parse_datetime(commit.get("committedDate")).astimezone(timezone.utc),
+        ),
+    )
+
+    return matching_file, blame
+
+
+def _make_ref_query(ref: str, blame_queries: str, index: int):
+    return f"""
+        ref{index}: ref(qualifiedName: "{ref}") {{
+            target {{
+                ... on Commit {{{blame_queries}
+                }}
+            }}
+        }}"""
+
+
+def _make_blame_query(path: str, index: int):
+    return f"""
+                    blame{index}: blame(path: "{path}") {{
```

Comment:
You'll notice that there is a lot of whitespace in these query fragments which makes it look a bit wonky here. I do this so that the query if formatted nicely and we can test it easily. With this format and how the test code does the comparison, we get really helpful diffs in case of a mismatch.

Reply:
Another note here is that bringing in a library to help with generating these GraphQL queries would be very useful. AFAIK there is nothing in the backend yet to help with this.

Reply:
Are you sure that is possible @NisanthanNanthakumar? I didn't think that graphql variables could pass arbitrary queries. I'm pretty sure they have to be typed and pass a value to a specific parameter. I did want to use variables here, but since I am making a dynamic number of repo/ref/path queries, I can't do it with variables alone unfortunately.